File: app/src/main/java/com/example/myapplication/event/OneActivity.java
package com.example.myapplication.event;

import android.animation.AnimatorSet;
import android.animation.ObjectAnimator;
import android.animation.ValueAnimator;
import android.annotation.SuppressLint;
import android.content.Intent;
import android.content.res.Resources;
import android.os.Bundle;
import android.util.Log;
import android.util.TypedValue;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewTreeObserver;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.example.myapplication.MainActivity;
import com.example.myapplication.R;
import com.example.myapplication.animator.ViewWrapper;
import com.example.myapplication.widget.CustomView;

public class OneActivity extends AppCompatActivity {

    private View customView;
    private int containerWidth;
    private int containerHeight;
    float lastX, lastY;
    RelativeLayout mRelativeLayout;

    @SuppressLint("ClickableViewAccessibility")
    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_one);
        customView = findViewById(R.id.btn);
        mRelativeLayout = findViewById(R.id.rl);
//        customView.setOnClickListener(new View.OnClickListener() {
//            @Override
//            public void onClick(View v) {
//                startActivity(new Intent(OneActivity.this,TwoActivity.class));
//            }
//        });

        ViewTreeObserver viewTreeObserver = customView.getViewTreeObserver();
        viewTreeObserver.addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
            @Override
            public void onGlobalLayout() {

            }
        });

//        customView.setOnTouchListener(new View.OnTouchListener() {
//
//            @Override
//            public boolean onTouch(View v, MotionEvent event) {
//                switch (event.getActionMasked()) {
//                    case MotionEvent.ACTION_DOWN:
//                        lastX = event.getRawX();
//                        lastY = event.getRawY();
//                        return true;
//                    case MotionEvent.ACTION_MOVE:
//                        //  不要直接用getX和getY,这两个获取的数据已经是经过处理的,容易出现图片抖动的情况
//                        float distanceX = lastX - event.getRawX();
//                        float distanceY = lastY - event.getRawY();
//
//                        float nextY = customView.getY() - distanceY;
//                        float nextX = customView.getX() - distanceX;
//
//                        // 不能移出屏幕
//                        if (nextY < 0) {
//                            nextY = 0;
//                        } else if (nextY > containerHeight - customView.getHeight()) {
//                            nextY = containerHeight - customView.getHeight();
//                        }
//                        if (nextX < 0) {
//                            nextX = 0;
//                        }else if (nextX > containerWidth - customView.getWidth()) {
//                            nextX = containerWidth - customView.getWidth();
//                        }
//                        // 属性动画移动
//                        ObjectAnimator y = ObjectAnimator.ofFloat(customView, "y", customView.getY(), nextY);
//                        ObjectAnimator x = ObjectAnimator.ofFloat(customView, "x", customView.getX(), nextX);
//
//                        x.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
//                            @Override
//                            public void onAnimationUpdate(ValueAnimator animation) {
//                                Log.e("OneActivity",animation.toString());
//                            }
//                        });
//
//                        AnimatorSet animatorSet = new AnimatorSet();
//                        animatorSet.playTogether(x, y);
//                        animatorSet.setDuration(0);
//                        animatorSet.start();
//
//                        lastX = event.getRawX();
//                        lastY = event.getRawY();
//                }
//                return false;
//            }
//        });

        customView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Log.e("OneActivity","View点击事件");
                Log.e("OneActivity","View宽度 ="+ customView.getWidth());

                // 包装类的属性动画 提供 get 和 set 方法
                ViewWrapper wrapper = new ViewWrapper(customView);
//                wrapper.setWidth(dp2px(80));
                ObjectAnimator.ofInt(wrapper,"width",dp2px(500))
                        .setDuration(1000).start();
                startActivity(new Intent(OneActivity.this,MainActivity.class));
            }
        });

//        ObjectAnimator.ofFloat(customView,"translationX",0f,300f).setDuration(2000).start();
    }


    @Override
    public void onWindowFocusChanged(boolean hasFocus) {
        super.onWindowFocusChanged(hasFocus);
        // 这里来获取容器的宽和高
        if (hasFocus) {
            containerHeight = mRelativeLayout.getHeight();
            containerWidth = mRelativeLayout.getWidth();
        }
    }

    public int dp2px(int dp) {
        return (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp, Resources.getSystem().getDisplayMetrics());
    }

    // 表示 Activity 的布局已经添加到 DecorView 中了
    @Override
    public void onContentChanged() {
        super.onContentChanged();
    }

}

File: app/src/main/java/com/example/myapplication/event/OneActivity.kt
//package com.example.myapplication.event
//
//import android.animation.ObjectAnimator
//import android.animation.PropertyValuesHolder
//import android.content.Intent
//import android.os.Bundle
//import android.util.Log
//import android.view.GestureDetector
//import android.view.MotionEvent
//import android.view.VelocityTracker
//import android.widget.Button
//import android.widget.TextView
//import androidx.appcompat.app.AppCompatActivity
//import com.example.myapplication.R
//import kotlinx.android.synthetic.main.activity_one.*
//
//
////class OneActivity : AppCompatActivity(),GestureDetector.OnGestureListener,GestureDetector.OnDoubleTapListener {
//class OneActivity : AppCompatActivity() {
//
//    override fun onCreate(savedInstanceState: Bundle?) {
//        super.onCreate(savedInstanceState)
//        setContentView(R.layout.activity_one)
//        EventBus.getInstance().register(this)
//        btn.setOnClickListener {
//            startActivity(Intent(this@OneActivity, TwoActivity::class.java))
//        }
//
////        ObjectAnimator.ofFloat(btn,"translationX",0f,500f).setDuration(2000).start()
//
////        initGestureDetector()
//    }
//
//    @Subscribe
//    fun eventMethod(a: Any) {
//        Log.e("OneActivity", a.toString())
//    }
//
//    override fun onDestroy() {
//        super.onDestroy()
//        EventBus.getInstance().unregister(this)
//
//        // 清除速度追踪并回收内存
////        velocityTracker.clear()
////        velocityTracker.recycle()
//    }
//
//
////    var lastX = 0f
////    var lastY = 0f
////    // 速度追踪
////    var velocityTracker = VelocityTracker.obtain()
////    override fun onTouchEvent(event: MotionEvent): Boolean {
////        var x = event.rawX;
////        var y = event.rawY;
////        velocityTracker.addMovement(event)
////        when (event.action) {
////            MotionEvent.ACTION_DOWN -> {
////                velocityTracker.computeCurrentVelocity(100)
////                val xVelocity = velocityTracker.xVelocity
////                Log.e("水平滑动速度 = ", xVelocity.toString())
////                val yVelocity = velocityTracker.yVelocity
////                Log.e("垂直滑动速度 = ", yVelocity.toString())
////            }
////            MotionEvent.ACTION_MOVE -> {
////                var deltaX = x - lastX
////                var deltaY = y - lastY
////
////            }
////            MotionEvent.ACTION_UP -> {
////                velocityTracker.computeCurrentVelocity(100)
////                val xVelocity = velocityTracker.xVelocity
////                Log.e("水平滑动速度 ACTION_UP = ", xVelocity.toString())
////                val yVelocity = velocityTracker.yVelocity
////                Log.e("垂直滑动速度 ACTION_UP = ", yVelocity.toString())
////            }
////        }
////
////        lastX = x
////        lastY = y
////
////        // 接管目标 View 的 onTouchEvent 方法
////        val gestureDetectorEvent = gestureDetector?.onTouchEvent(event)
//////        return gestureDetectorEvent ?: false
//////        return super.onTouchEvent(event)
////        return true
////    }
//
//    // 手势检测
////    var gestureDetector: GestureDetector? = null
////    fun initGestureDetector(){
////        gestureDetector = GestureDetector(this,this)
////        // 解决长按屏幕后无法拖动的现象
////        gestureDetector?.setIsLongpressEnabled(false)
////    }
////
////    // 手指轻轻触摸屏幕的一瞬间
////    override fun onDown(e: MotionEvent?): Boolean {
////        return false
////    }
////
////    // 手指轻轻触摸屏幕，尚未松开或拖动，这里区分 onDown 方法
////    override fun onShowPress(e: MotionEvent?) {
////    }
////
////    // 手指轻触屏幕后松开，单击行为
////    override fun onSingleTapUp(e: MotionEvent?): Boolean {
////        return false
////    }
////
////    // 手指按下屏幕并拖动，这是拖动行为
////    override fun onScroll(
////        e1: MotionEvent?,
////        e2: MotionEvent?,
////        distanceX: Float,
////        distanceY: Float
////    ): Boolean {
////        return false
////    }
////
////    // 长按事件
////    override fun onLongPress(e: MotionEvent?) {
////
////    }
////
////    // 按下触摸屏，快速滑动后松开，快速滑动行为
////    override fun onFling(e1: MotionEvent?,e2: MotionEvent?,velocityX: Float,velocityY: Float): Boolean {
////        return false
////    }
////
////    // 双击，由连续两次的单击组成
////    override fun onDoubleTap(e: MotionEvent?): Boolean {
////        return false
////    }
////
////    // 表示发生了双击行为
////    override fun onDoubleTapEvent(e: MotionEvent?): Boolean {
////        return false
////    }
////
////    // 严格的单击行为
////    override fun onSingleTapConfirmed(e: MotionEvent?): Boolean {
////        return false
////    }
//}
